### PR TITLE
Updated SigV4 signing library in Gremlin connection for Neptune connector

### PR DESCRIPTION
**Pull request by xiazcy**
*Issue #, if available:*
#1574

*Description of changes:*
The SigV4 library used for IAM in Gremlin connection in Neptune connector is a deprecated library, this PR updates it to use the new maintained library for signing. The deprecated library is only recommended to be used with 3.4.x Gremlin drivers, and may work when using 3.5.x and earlier 3.6.x driver versions, but will no longer work with 3.7.x.

Also downgraded the Gremlin driver version used by the connector to within the supported range for Neptune (currently 3.6.5 is the max). 

I was able to reproduce the error on 3.7.1 gremlin console, trying to connect to IAM enabled cluster using the old SigV4 signing library:
image

I have tested the fix in this PR by deploying a customer connector and querying an IAM enabled cluster, and was successfull:
image
Screenshot 2024-01-19 at 5 26 13 PM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by xiazcy** [comment]
This PR also provides a fix to the custom schema writer that allows a workaround for issue https://github.com/awslabs/aws-athena-query-federation/issues/1655, where naming clashes can be avoided by creating a view from [custom queries](https://github.com/awslabs/aws-athena-query-federation/tree/master/athena-neptune/docs/aws-glue-sample-scripts#custom-query).

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/rowwriters/CustomSchemaRowWriter.java [hunk 1]
```diff
@@ -105,20 +113,26 @@ public static void writeRowTemplate(RowWriterBuilder rowWriterBuilder, Field fie
...
+                                logger.debug("writeRowTemplate VARCHAR*" + field.getName() + "*" + minorType + "*"
+                                        + (fieldValue == null ? "" : fieldValue.getClass()) + "*");
 
                                 if (fieldValue != null) {
                                     if (fieldValue.getClass().equals(String.class)) {
                                         value.value = fieldValue.toString();
                                         value.isSet = 1;
                                     }
-                                    else {       
+                                    else if (fieldValue instanceof ArrayList) {
```
Comment (chngpe) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/rowwriters/CustomSchemaRowWriter.java:124]: Do we need to apply this condition `else if (fieldValue instanceof ArrayList)` to rest of types?  For example `DATEMILLI`

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/rowwriters/CustomSchemaRowWriter.java [hunk 2]
```diff
@@ -197,11 +217,13 @@ public static void writeRowTemplate(RowWriterBuilder rowWriterBuilder, Field fie
...
+                                    + (fieldValue == null ? "" : fieldValue.getClass()) + "*");
+                            if (fieldValue.getClass().equals(Float.class)) {
                                 value.value = Float.parseFloat(fieldValue.toString());
                                 value.isSet = 1;
                             }
-                            else { 
+                            else if (fieldValue instanceof ArrayList) {
                                 ArrayList<Object> objValues = (ArrayList) fieldValue;
                                 if (objValues != null && objValues.get(0) != null && !(objValues.get(0).toString().trim().isEmpty())) {
                                     value.value = Float.parseFloat(objValues.get(0).toString());
```
[the remark this answers is not in the document]
Comment (xiazcy) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/rowwriters/CustomSchemaRowWriter.java:229]: We can only return zero or one value for a column. Custom query may return a single value wrapped inside a list/set, so we needed to access the element. In corner cases multiple values are returned, we are also only returning the first value based on this constraint.

**Comment by triggan** [comment]
@chngpe - Gentle ping to see if we can get this reviewed and merged soon.  This is blocking a few customers as the current version of the connector does not work with an IAM-enabled Neptune cluster.

**Review by chngpe — APPROVED** [review]

**Comment by xiazcy** [comment]
Hi @chngpe, is there an ETA for merging this PR? Or is there anything else I should update? Thanks!

**Comment by AbdulR3hman** [comment]
@xiazcy done; thank you for your contribution.

**Pull request merged**